File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    Integer prev = null;  // Store the previous value during in-order traversal

    public boolean isValidBST(TreeNode root) {
        return inOrderTraversal(root);
    }

    private boolean inOrderTraversal(TreeNode node) {
        if (node == null) return true;  // Base case: Empty tree is a BST

        // Left subtree check
        if (!inOrderTraversal(node.left)) return false;

        // Check current node against previous node
        if (prev != null && node.val <= prev) return false;

        prev = node.val;  // Update previous node value

        // Right subtree check
        return inOrderTraversal(node.right);
    }
}


// public class Solution {

//     List<Integer> list = new LinkedList<>();

//     public boolean isValidBST(TreeNode root) {
//        inOrderTraversal(root);

//        if ( list == null ) return true;
//        for ( int i = 0; i < list.size() - 1; i++) {
//         int curr = list.get(i);
//         int next = list.get(i+1);
//         if ( curr >= next ) return false;
//        }
//        return true;
//     }

//     private void inOrderTraversal(TreeNode root) {
//         // Base case
//         if ( root == null ) return;

//         // Recursive case
//         inOrderTraversal(root.left);
//         list.add(root.val);
//         inOrderTraversal(root.right);
//     }
// }

File: Data Structures & Algorithms/climbing-stairs/submission-5.py
class Solution:
    def climbStairs(self, n: int) -> int:

        if n <= 2:
            return n

        dp = [0] * (n + 1)
        dp[1] = 1
        dp[2] = 2

        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]

        return dp[n]


# class Solution:
#     def climbStairs(self, n: int) -> int:

#         cache = [-1] * n
#         def dfs(i):
            
#             # Base Case
#             if i >= n:
#                 return i == n

#             # Caching check
#             if cache[i] != -1:
#                 return cache[i]

#             # Recursive Case
#             cache[i] =  dfs(i + 1) + dfs(i + 2)
#             return cache[i]

#         return dfs(0)
        